Search suffix tree implementation

hannojg opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Explanation of Change
<!-- Explain what your change does and how it addresses the linked issue -->

Reapplying the changes from this reverted PR:

- https://github.com/Expensify/App/pull/48652

This PR makes sure to fix the following regressions as well:

- [x] https://github.com/Expensify/App/pull/51332#issuecomment-2432573981
- [x] https://github.com/Expensify/App/issues/51471#issuecomment-2442274872
- [x] ~~https://github.com/Expensify/App/issues/51123~~ 
	- ([Turns out this wasn't caused by these changes](https://github.com/Expensify/App/issues/51123#issuecomment-2456513922))


### Fixed Issues
<!---
1. Please postfix `$` with a URL link to the GitHub issue this Pull Request is fixing. For example, `$ https://github.com/Expensify/App/issues/<issueID>`.
2. Please postfix  `PROPOSAL:` with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).  For example, `PROPOSAL: https://github.com/Expensify/App/issues/<issueID>#issuecomment-1369752925`

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking and its automation will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<issueID>
$ https://github.com/Expensify/App/issues/<issueID(comment)>

Do NOT only link the issue number like this: $ #<issueID>
--->
$ https://github.com/Expensify/App/issues/46591
PROPOSAL:


### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

**Test 1:**
- Open the chat finder page (search icon on the start page when authenticated)
- Perform searches, confirm that everything is found, make sure to also search for numbers
- Search for an email address (e.g. `hanno@margelo.io`), make sure it's found. Then search for `hannomargeloio` (removing all special chars), and make sure its still found
- You might want to open the production version and compare your search results with the dev version to make sure you got the similar results (see note at the top)

**Test 2:**
- Search for the user's own email address, make sure that self DM results are at the top (if its a fresh user create a few threads)

**Test 3:**
- Create a new group with three participants
- Search for the group name, including the commas, confirm that it shows in the search results

### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->

Same as testing steps

### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

Same as testing steps

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

<!-- add screenshots or videos here -->


https://github.com/user-attachments/assets/57bca41b-6f56-4f5a-ad23-bf25ba9465d5



</details>

<details>
<summary>Android: mWeb Chrome</summary>

<!-- add screenshots or videos here -->

https://github.com/user-attachments/assets/2c62b027-6d00-42f3-825d-28460cdbd95f


</details>

<details>
<summary>iOS: Native</summary>

<!-- add screenshots or videos here -->


https://github.com/user-attachments/assets/1102d13d-6bdf-41f8-b319-cff930399c71



</details>

<details>
<summary>iOS: mWeb Safari</summary>

<!-- add screenshots or videos here -->


https://github.com/user-attachments/assets/94f08c09-e0c2-479c-a64f-77cf40f5c8f9



</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

<!-- add screenshots or videos here -->


https://github.com/user-attachments/assets/fabbdaac-005d-4e74-8fec-72c8dac31e16



</details>

<details>
<summary>MacOS: Desktop</summary>

<!-- add screenshots or videos here -->

https://github.com/user-attachments/assets/597b8c09-9406-4aa4-90c9-18c9a7af7b00

</details>

hannojg committed:
exclude comma from search values

hannojg committed:
Revert "Revert "Revert "Revert "Search suffix tree implementation""""

This reverts commit 20af9f625b5e636f3e2cce928d4eb5ab5954f8ee.

hannojg committed:
wip: refactoring test to be reusable

hannojg committed:
Revert "wip: refactoring test to be reusable"

This reverts commit a2d8012769451cccaea1e27f2ddf4154abcad201.

hannojg committed:
Merge branch 'main' of github.com:Expensify/App into perf/search-tree

hannojg committed:
fix: sort search results correctly

ahmedGaber93 commented:
BUG: search result display previous search result above the current search result 
1. login with high traffic account
2. search for `marc`
3. search for `shawn`

Expected: search for `shawn` should display result for `shawn` only
Current: it displays result for `marc` and `shawn`

<details>
<summary>screenshots</summary>

![Screenshot 2024-11-06 at 4 42 46 PM](https://github.com/user-attachments/assets/50fd982b-1bd4-4f9d-8e63-aa2708d8477f)

![Screenshot 2024-11-06 at 4 41 55 PM](https://github.com/user-attachments/assets/f93301ec-3216-48a4-9201-e5f43e5d85d6)


</details>
<br>

BUG: Ordering result prefer `contain` over `startWith` 


<details>
<summary>screenshots</summary>

![Screenshot 2024-11-06 at 4 42 46 PM](https://github.com/user-attachments/assets/50fd982b-1bd4-4f9d-8e63-aa2708d8477f)

![Screenshot 2024-11-06 at 4 41 55 PM](https://github.com/user-attachments/assets/f93301ec-3216-48a4-9201-e5f43e5d85d6)


</details>

@hannojg

hannojg commented:
Thanks for testing, let me fix those!

hannojg committed:
Merge branch 'main' of github.com:Expensify/App into perf/search-tree

hannojg committed:
cleanup option list

hannojg committed:
fix duplicate search results

hannojg commented:
I split out a few changes here first:

- https://github.com/Expensify/App/pull/52168

hannojg committed:
Merge branch 'main' of github.com:Expensify/App into perf/search-tree

hannojg committed:
eslint

ahmedGaber93 commented:
Status of fixed issues:
- [x] BUG: search result display previous search result above the current search result
- [ ] BUG: Ordering result prefer includes over startsWith
- [x] BUG: Duplicated result

hannojg committed:
wip

hannojg commented:
Thanks for testing! On the ordering bug, i looked into it and came to the conclusion that i first need to cleanup the OptionListUtils a little, its a lot of spaghetti 🍝 

How it currently works is:

1. (`getOptions`) first convert all reports and users into a different format (the SearchOption format)
2. (`getOptions`) then order all search options
3. (`filterOptions`) then filter them
4. (`filterOptions`) then apply a second ordering (i think this double ordering we can maybe get rid of in the cleanup)

All of this is happening in like 2 functions and is very coupled.

The reason why the ordering is broken with our SuffixTree is that right now it works like this:

1. (`getOptions`) first convert all reports and users into a different format (the SearchOption format)
2. (`getOptions`) then order all search options
3. BREAKING(`FastSearch`): filter options using our search tree - the search tree doesn't keep the order in-tact, so its breaking here
4. (`FastSearch`) apply the second ordering

My current approach is to cleanup the current code so the search works like this (before applying the changes from this PR):

1. convert all to SearchOption (unfortunately we have to do this first for multiple reasons... will see if i can make a bigger refactor so that only search results that we currently need to display will be converted which would be better for performance)
2. filter them
3. then order them at once

This way the ordering should always give the same results. Also it should be more performant to just filter the limited search results, instead of ordering everything at the start.

ahmedGaber93 commented:
Great! We can wait cleanup PRs then retest.

grgia commented on src/libs/FastSearch.ts:
```diff
@@ -0,0 +1,140 @@
+};
+
+// There are certain characters appear very often in our search data (email addresses), which we don't need to search for.
+const charSetToSkip = new Set(['@', '.', '#', '$', '%', '&', '*', '+', '-', '/', ':', ';', '<', '=', '>', '?', '_', '~', '!', ' ', ',']);
```
should this be a CONST somewhere/do we forsee this list growing?

hannojg replied:
hm, not really sure if this list would grow or be used anywhere else 🤔

marcaaron commented:
Is this PR on HOLD for https://github.com/Expensify/App/pull/52250?

ahmedGaber93 commented:
I think https://github.com/Expensify/App/pull/52250 is the only one left on cleanup PRs, then @hannojg will add some changes here to fix the ordering as I think.

marcaaron commented:
Seems we are almost ready to continue now that the linked PR has been merged?

hannojg commented:
Yes, will continue now. A few more cleanups / restructures are needed before but we are getting closer!!

hannojg commented:
Next cleanup PR is available for review:

- https://github.com/Expensify/App/pull/53056

hannojg commented:
Working on the final cleanup PR that will provide a clear separation between filtering and ordering here:

- https://github.com/Expensify/App/pull/53371

(should be good in a few hours, will update here once done)

hannojg commented:
Alright, all cleanups have been landed - working now on fixing this one!

hannojg committed:
Merge branch 'main' of github.com:Expensify/App into perf/search-tree

hannojg committed:
fixes after merge

hannojg committed:
wip: use fast search in SearchRouterList

hannojg committed:
cleanup tests

hannojg committed:
add `useFastSearchFromOptions` hook

hannojg committed:
remove comment

hannojg committed:
remove unnecessary test case

hannojg committed:
add docs

hannojg committed:
remove obsolete test

hannojg commented:
@ahmedGaber93 @marcaaron ready for review 😊

ahmedGaber93 commented:
Bug: duplicate result on `SearchRouterList`

https://github.com/user-attachments/assets/d7c192c6-aa0f-4ec9-b08e-8ea1612b9c4e

ahmedGaber93 commented:
The search result ordering is Different between DEV and Production, is this behavior expected?

ahmedGaber93 commented:
the `searchTerms` looks not supported, are we intended to support it on this PR?  
e.g. search by "ah ga" should return result like "ahmed gaber" `(ah)med (ga)ber`




https://github.com/user-attachments/assets/a1423c3e-359c-41c1-a905-8b1e761f6653

hannojg committed:
refactor out `combinedOrderingOfReportsAndPersonalDetails`

hannojg committed:
fix option A

hannojg committed:
different paths for sortByReportTypeInSearch true and false

hannojg commented:
> the searchTerms looks not supported, are we intended to support it on this PR?

Thats a good observation and something I've missed noticing so far. I need to think about how this can efficiently be added to the current implementation!

The other two problems you mentioned should be fixed though!

marcaaron commented on src/libs/OptionsListUtils.ts:
```diff
@@ -1730,45 +1746,55 @@ function filterOptions(options: Options, searchInputValue: string, config?: Filt
-    let filterResult = options;
-    if (searchInputValue.trim().length > 0) {
-        filterResult = filterOptions(options, searchInputValue, config);
+function combinedOrderingOfReportsAndPersonalDetails(
```
```suggestion
function combineOrderingOfReportsAndPersonalDetails(
```

marcaaron reviewed (changes requested):
Looks good overall.

hannojg committed:
add functionality to filter by sub words

hannojg committed:
rename function to combineOrderingOfReportsAndPersonalDetails

hannojg committed:
make config optional

hannojg committed:
add brackets to ignore list

hannojg committed:
add test

hannojg commented:
> the searchTerms looks not supported, are we intended to support it on this PR?

Added support for this as well now + added a test! 😊

hannojg committed:
Merge branch 'main' of github.com:Expensify/App into perf/search-tree

hannojg commented:
Raised a question [here on slack](https://margelo.slack.com/archives/C01GTK53T8Q/p1734431879247749) about the failing eslint tests

hannojg committed:
no default ID for strings

hannojg committed:
add to large files

hannojg commented:
As discussed [here](https://margelo.slack.com/archives/C01GTK53T8Q/p1734435714071049?thread_ts=1734427537.784129&cid=C01GTK53T8Q) `OptionsListUtils` will be migrated separately

ahmedGaber93 commented:
## Reviewer Checklist

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android: Native
    - [ ] Android: mWeb Chrome
    - [ ] iOS: Native
    - [ ] iOS: mWeb Safari
    - [ ] MacOS: Chrome / Safari
    - [ ] MacOS: Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.ts or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [ ] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [ ] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [ ] I verified that all the inputs inside a form are aligned with each other.
    - [ ] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] For any bug fix or new feature in this PR, I verified that sufficient [unit tests](https://github.com/Expensify/App/blob/main/tests/README.md) are included to prevent regressions in this flow.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Android: mWeb Chrome</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS: Native</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS: mWeb Safari</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>MacOS: Desktop</summary>

<!-- add screenshots or videos here -->

</details>

ahmedGaber93 commented:
Thanks for the updates, I think userToInvite have been broken in the last changes


https://github.com/user-attachments/assets/59642a7f-1bd3-444d-8708-3aaefa09499d

hannojg committed:
Merge branch 'main' of github.com:Expensify/App into perf/search-tree

hannojg committed:
Merge branch 'main' of github.com:Expensify/App into perf/search-tree

hannojg committed:
fix duplicated personal details https://github.com/Expensify/App/issues/53579

This is a more effective implementation for https://github.com/Expensify/App/pull/54147/files which i reverted in this PR

hannojg committed:
add changes from https://github.com/Expensify/App/pull/54147 back

hannojg committed:
add functionality back to check if current option matches search or not

hannojg committed:
pass filtered options

hannojg committed:
typo

hannojg committed:
fix up tests

hannojg committed:
remove debug line

hannojg commented:
This should be good again for testing @ahmedGaber93

ahmedGaber93 reviewed (approved):
Tested well, All yours @marcaaron

marcaaron commented on src/hooks/useFastSearchFromOptions.ts:
````diff
@@ -0,0 +1,113 @@
+ * Builds a suffix tree and returns a function to search in it.
+ *
+ * @example
+ * ```
````
NAB
```suggestion
 *
```

marcaaron commented on src/hooks/useFastSearchFromOptions.ts:
```diff
@@ -0,0 +1,113 @@
+            }
+
+            // The user might separated words with spaces to do a search such as: "jo d" -> "john doe"
+            // With the suffix search tree you can only search for one word at a time. Its most efficient to search for the longest word,
```
NAB
```suggestion
            // With the suffix search tree, you can only search for one word at a time. It is most efficient to search for the longest word
```
